pkg/controller/cluster/cluster_controller.go

```diff
@@ -151,20 +204,99 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 	}
 
-	klog.Infof("reconciling cluster object %v triggers idempotent reconcile.", name)
+	klog.Infof("Reconciling cluster object %v triggers idempotent reconcile.", name)
 	if err := r.actuator.Reconcile(cluster); err != nil {
 		if requeueErr, ok := errors.Cause(err).(controllerError.HasRequeueAfterError); ok {
 			klog.Infof("Actuator returned requeue-after error: %v", requeueErr)
 			return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
 		}
 		klog.Errorf("Error reconciling cluster object %v; %v", name, err)
-		return reconcile.Result{}, err
+		return reconcile.Result{}, errors.WithStack(err)
 	}
 	return reconcile.Result{}, nil
 }
+
+// listChildren returns a list of Deployments, Sets, and Machines than have an ownerref to the given cluster
+func (r *ReconcileCluster) listChildren(ctx context.Context, cluster *clusterv1.Cluster) ([]runtime.Object, error) {
+	var children []runtime.Object
+
+	ns := cluster.GetNamespace()
+	opts := metav1.ListOptions{
+		LabelSelector: labels.FormatLabels(
+			map[string]string{clusterv1.MachineClusterLabelName: cluster.GetName()},
+		),
+	}
+
+	dfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineDeployments(ns).List(m)
+	}
+	sfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.MachineSets(ns).List(m)
+	}
+	mfunc := func(ctx context.Context, m metav1.ListOptions) (runtime.Object, error) {
+		return r.clusterClient.Machines(ns).List(m)
+	}
+
+	deployments, err := pager.New(dfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "failed to list MachineDeployments in %s for cluster %s", ns, cluster.Name)
+	}
+	dlist, ok := deployments.(*clusterv1.MachineDeploymentList)
+	if !ok {
+		return []runtime.Object{}, fmt.Errorf("Expected MachineDeploymentList, got %T", deployments)
+	}
+
+	sets, err := pager.New(sfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "Failed to list MachineSets in %s", ns)
+	}
+	slist, ok := sets.(*clusterv1.MachineSetList)
+	if !ok {
+		return []runtime.Object{}, fmt.Errorf("Expected MachineSetList, got %T", sets)
+	}
+
+	machines, err := pager.New(mfunc).List(ctx, opts)
+	if err != nil {
+		return []runtime.Object{}, errors.Wrapf(err, "Failed to list MachineSets in %s", ns)
+	}
+	mlist, ok := machines.(*clusterv1.MachineList)
+	if !ok {
+		return []runtime.Object{}, fmt.Errorf("Expected MachineList, got %T", machines)
+	}
+
+	for _, d := range dlist.Items {
+		if pointsTo(&d.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, d.DeepCopyObject())
+		}
+	}
+
+	for _, s := range slist.Items {
+		if pointsTo(&s.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, s.DeepCopyObject())
+		}
+	}
+
+	for _, m := range mlist.Items {
+		if pointsTo(&m.ObjectMeta, &cluster.ObjectMeta) {
+			children = append(children, m.DeepCopyObject())
```

Comment:
I wonder if we should only append Machines if they do **not** belong to a MachineSet?